src/types.ts

```diff
@@ -0,0 +1,182 @@
+import { Cookie, Headers, HttpMethod, Page, ResourceType, Viewport } from 'puppeteer'
+// import { Viewport } from 'puppeteer/DeviceDescriptors'
+
+interface IAuthlessCore {
+  name: string
+  domainPathRouter: IDomainPathRouter
+}
+
+interface IBotRouter {
+  botMap: {[url: string ]: IBot[]}
+  botIndices: {[url: string ]: number}
+  getBotForUrl(url: string)
+  // getBot(): IBot | undefined
+  getBotByUsername(username: string): IBot | undefined
+}
+
+interface IBot {
+  username: string
+  password: string
+  hitCount: number
+  loginCount: number
+  captchaCount: number
+  foundLogin: (found: Boolean) => void
+  foundCaptcha: (found: Boolean) => void
+  getHitCount: () => number
+  getLoginHitCount: () => number
+  getCaptchaHitCount: () => number
+}
+
+interface IResponseMeta {
+  url: string
+  viewport: Viewport
+  title: string
+}
+
+interface IResponse {
+  meta: IResponseMeta
+  page: any
+  content: any
+  cookies: Cookie[]
+  xhrs: Xhr[]
+}
+
+interface IDomainPathRouter {
+  domainMap: {[url: string ]: IDomainPath}
+  getDomainPathForUrl: (url: string) => IDomainPath | undefined
+}
+
+interface IDomainPath {
+  domain: string
+  // urls: string[]
+  // botRouter: IBotRouter
+  // isAuthenticated: (page: any) => Promise<Boolean>
+  // authenticate: (page: any, bot?: IBot) => Promise<Boolean | string>
+  pageHandler: (page: Page, bot?: IBot, config?: any) => Promise<IResponse | null>
```

Comment:
1. Why is it called `pageHandler`? Thinking through the API surface, is there a particular reason why we don't call it `fetch` as well?

1. When/why does it return `null`?

Reply:
I guess because it doesn't fetch any data. Any data which comes in is saved into `this.responses` with a listener, so this page doesn't have to return anything.
It is analogous to a user interacting with the page, the data gets saved in the background.
When interaction is done, we save the data we received and the contents of the page at the end.